src/mbgl/style/expression/type_check.cpp

```diff
@@ -0,0 +1,121 @@
+#include <mbgl/style/expression/type_check.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using namespace mbgl::style::expression::type;
+
+TypecheckResult typecheck(const Type& expected, const std::unique_ptr<Expression>& e) {
```

Comment:
> In C++, you could just provide the evaluation methods, and the signatures could be deduced from those.

Hm, I don't see how this would work...  I think we need the list of signatures to be modeled as a value (e.g. `std::vector<std::vector<type::Type>>`) in order to do type checking.  That wouldn't be possible just from the existence of certain evaluation methods, right?